torch_geometric/nn/models/GraphMask.py

```diff
@@ -0,0 +1,839 @@
+            latest_target_embeddings = []
+            for module in self.model.modules():
+                if isinstance(module, MessagePassing):
+                    latest_source_embeddings.append(
+                        module.latest_source_embeddings)
+                    latest_messages.append(module.latest_messages)
+                    latest_target_embeddings.append(
+                        module.latest_target_embeddings)
+            gate_input = [
+                latest_source_embeddings, latest_messages,
+                latest_target_embeddings
+            ]
+            if self.log:  # pragma: no cover
+                pbar = tqdm(total=self.num_layers)
+            for i in range(self.num_layers):
+                if self.log:  # pragma: no cover
+                    if self.allow_multiple_explanations:
+                        pbar.set_description(f'Explain nodes {node_idx}')
+                    else:
+                        pbar.set_description(f'Explain node {node_idx}')
+                output = self.full_biases[i]
+                for j in range(len(gate_input)):
+                    partial = self.gates[i * 4][j](gate_input[j][i])
+                    result = self.gates[(i * 4) + 1][j](partial)
+                    output = output + result
+                relu_output = self.gates[(i * 4) + 2](output / len(gate_input))
+                sampling_weights = self.gates[(i * 4) +
+                                              3](relu_output).squeeze(dim=-1)
+                sampling_weights, _ = self.__hard_concrete__(
+                    sampling_weights, training=False)
+                if i == 0:
+                    edge_weight = sampling_weights
+                else:
+                    if (edge_weight.size(-1) != sampling_weights.size(-1)
+                            and self.layer_type == 'GAT'):
+                        sampling_weights = F.pad(
+                            input=sampling_weights,
+                            pad=(0, edge_weight.size(-1) -
+                                 sampling_weights.size(-1), 0, 0),
+                            mode='constant', value=0)
+                    edge_weight = torch.cat((edge_weight, sampling_weights), 0)
+                if self.log:  # pragma: no cover
+                    pbar.update(1)
+        if self.log:  # pragma: no cover
+            pbar.close()
+
+        edge_mask = edge_weight.view(-1,
+                                     edge_weight.size(0) // self.num_layers)
+        edge_mask = torch.mean(edge_mask, 0)
+
+        return edge_mask
+
+    def __train_graph_explainer__(self, graph_idx, x, edge_index, **kwargs):
+        r"""Learns a node feature mask and an edge mask and returns only the
+        learned node feature mask that plays a crucial role to explain the
+        prediction made by the GNN for graph :attr:`graph_idx`.
+
+        Args:
+            graph_idx (int): Graph to explain.
```

Comment:
Doesn't seem to be used below?

Reply:
This `graph_idx` helps letting the user know which graph's predictions are to be explained. Hope, this helps broadening your understanding a bit!!